File: src/converter/plugins/DeclarationMergingPlugin.ts
import ts, {
    Declaration, HeritageClause, InterfaceDeclaration,
    ModuleDeclaration,
    NamedDeclaration,
    Program,
    SignatureDeclaration,
    SymbolTable, TypeParameterDeclaration
} from "typescript";
import {ConverterPlugin} from "../plugin.js";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {DeepMap} from "../../utils/deepMap.js";
import {GeneratedFile} from "../generated.js";
import {NamespaceStrategy} from "../../configuration/configuration.js";

export const declarationMergingServiceKey = Symbol()

export class DeclarationMergingService {
    private readonly virtualSourceFile = ts.createSourceFile("virtual.d.ts", "", ts.ScriptTarget.Latest)
    private readonly printer = ts.createPrinter({
        removeComments: true,
        newLine: ts.NewLineKind.LineFeed,
    })

    constructor(public readonly program: Program) {
    }

    private coveredSymbols = new Set<ts.Symbol>()

    isCovered(node: NamedDeclaration): boolean {
        const symbol = this.getSymbol(node)
        if (!symbol) return false

        if (this.coveredSymbols.has(symbol)) return true

        const valueDeclaration = symbol.valueDeclaration

        // it is one of the additional declarations, and it will be covered in value declaration
        return valueDeclaration !== undefined
            && ts.isClassDeclaration(valueDeclaration)
            && node !== valueDeclaration;
    }

    cover(node: NamedDeclaration) {
        const symbol = this.getSymbol(node)
        if (!symbol) return

        this.coveredSymbols.add(symbol)
    }

    getTypeParameters(node: NamedDeclaration): TypeParameterDeclaration[] | undefined {
        const symbol = this.getSymbol(node)
        if (!symbol) return undefined

        const declarationTypeParameters = symbol.declarations
            ?.filter((it): it is InterfaceDeclaration => ts.isInterfaceDeclaration(it))
            ?.map(it => Array.from(it.typeParameters ?? []))

        const valueDeclarationTypeParameters =
            symbol.valueDeclaration !== undefined && ts.isClassDeclaration(symbol.valueDeclaration)
                ? Array.from(symbol.valueDeclaration.typeParameters ?? [])
                : undefined

        return [
            ...(declarationTypeParameters ?? []),
            valueDeclarationTypeParameters ?? [],
        ].reduce((typeParameters, result) => (
            typeParameters.length > result.length
                ? typeParameters
                : result
        ))
    }

    getHeritageClauses(node: NamedDeclaration): HeritageClause[] | undefined {
        const symbol = this.getSymbol(node)
        if (!symbol) return undefined

        const heritageClauses = symbol.declarations
                ?.filter((it): it is InterfaceDeclaration => ts.isInterfaceDeclaration(it))
                ?.flatMap(it => it.heritageClauses)
                ?.filter((it): it is HeritageClause => Boolean(it))

        if (ts.isClassDeclaration(node)) {
            const mainHeritageClauses = node.heritageClauses

            return [
                ...(mainHeritageClauses ?? []),
                ...(heritageClauses ?? []),
            ]
        } else {
            return heritageClauses
        }
    }

    getMembers(
        node: NamedDeclaration,
        resolveNamespaceStrategy: ((node: ModuleDeclaration) => NamespaceStrategy) | undefined,
    ): Declaration[] | undefined {
        const symbol = this.getSymbol(node)
        if (!symbol) return undefined

        const members = this.getUniqMembers(symbol.members)
            .filter(member => !ts.isTypeParameterDeclaration(member))

        const exports = this.getUniqMembers(symbol.exports)
            .filter(member => !ts.isTypeParameterDeclaration(member))

        return [...members, ...exports]
            .filter(member => {
                if (
                    ts.isModuleBlock(member.parent)
                    && resolveNamespaceStrategy?.(member.parent.parent) === "package"
                ) {
                    return false
                }

                if (
                    ts.isVariableDeclarationList(member.parent)
                    && ts.isVariableStatement(member.parent.parent)
                    && ts.isModuleBlock(member.parent.parent.parent)
                    && resolveNamespaceStrategy?.(member.parent.parent.parent.parent) === "package"
                ) {
                    return false
                }

                return true
            })
    }

    private getUniqMembers(symbolTable: SymbolTable | undefined): Declaration[] {
        const typeChecker = this.program.getTypeChecker()

        return Array.from(symbolTable?.values() ?? [])
            .flatMap(symbol => {
                const declarations = symbol.declarations ?? []
                const [firstDeclaration] = declarations

                if (!firstDeclaration) return []
                if (!this.isSignatureDeclaration(firstDeclaration)) return [firstDeclaration]

                const declarationMap = declarations.reduce((acc, declaration) => {
                    if (!this.isSignatureDeclaration(declaration)) return acc

                    const parameterSymbols = declaration.parameters
                        .map(parameter => {
                            const typeNode = parameter.type
                            if (!typeNode) return "any"

                            const type = typeChecker.getTypeFromTypeNode(typeNode)

                            return type?.symbol ?? this.printNode(typeNode)
                        })

                    acc.set(parameterSymbols, declaration)

                    return acc
                }, new DeepMap<ts.Symbol | string, Declaration>())

                return Array.from(declarationMap.values())
            })
    }

    private isSignatureDeclaration(declaration: Declaration): declaration is SignatureDeclaration {
        return ts.isCallSignatureDeclaration(declaration)
            || ts.isConstructSignatureDeclaration(declaration)
            || ts.isConstructorDeclaration(declaration)
            || ts.isMethodSignature(declaration)
            || ts.isMethodDeclaration(declaration)
    }

    private getSymbol(node: NamedDeclaration): ts.Symbol | undefined {
        if (!node.name) return undefined

        const typeChecker = this.program.getTypeChecker()
        return typeChecker.getSymbolAtLocation(node.name)
    }

    private printNode(node: ts.Node) {
        const sourceFile = node.getSourceFile() ?? this.virtualSourceFile

        return this.printer.printNode(ts.EmitHint.Unspecified, node, sourceFile)
    }
}

export class DeclarationMergingPlugin implements ConverterPlugin {
    private readonly declarationMergingService: DeclarationMergingService

    constructor(program: Program) {
        this.declarationMergingService = new DeclarationMergingService(program);
    }

    setup(context: ConverterContext): void {
        context.registerService(declarationMergingServiceKey, this.declarationMergingService)
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    generate(): GeneratedFile[] {
        return [];
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/DeclarationMergingPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.util.DeepMap
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.JsArrays
import js.array.ReadonlyArray
import js.objects.unsafeJso
import typescript.*
import kotlin.contracts.contract

@JsExport
val declarationMergingServiceKey = ContextKey<DeclarationMergingService>()

@JsExport
class DeclarationMergingService @JsExport.Ignore constructor(private val program: Program) {
    private val coveredSymbols = mutableSetOf<typescript.Symbol>()
    private val virtualSourceFile = createSourceFile("virtual.d.ts", "", ScriptTarget.Latest)
    private val printer = createPrinter(unsafeJso {
        removeComments = true
        newLine = NewLineKind.LineFeed
    })

    fun isCovered(node: NamedDeclaration): Boolean {
        val symbol = this.getSymbol(node)
        if (symbol == null) return false

        if (symbol in this.coveredSymbols) return true

        val valueDeclaration = symbol.valueDeclaration

        // it is one of the additional declarations, and it will be covered in value declaration
        return valueDeclaration != null
            && isClassDeclaration(valueDeclaration)
            && node !== valueDeclaration
    }

    fun cover(node: NamedDeclaration) {
        val symbol = this.getSymbol(node) ?: return

        this.coveredSymbols.add(symbol)
    }

    fun getTypeParameters(node: NamedDeclaration): ReadonlyArray<TypeParameterDeclaration>? {
        val symbol = this.getSymbol(node)
        if (symbol == null) return null

        val declarationTypeParameters = symbol.declarations
            ?.mapNotNull {
                if (!isInterfaceDeclaration(it)) return@mapNotNull null
                it.typeParameters?.asArray()
            }

        val valueDeclaration = symbol.valueDeclaration
        val valueDeclarationTypeParameters =
            if (valueDeclaration != null && isClassDeclaration(valueDeclaration)) {
                valueDeclaration.typeParameters?.asArray()
            } else {
                null
            }

        val typeParameters =
            (declarationTypeParameters ?: emptyList()) +
                    listOf(valueDeclarationTypeParameters ?: emptyArray())

        return typeParameters.maxBy { it.size }
    }

    fun getHeritageClauses(node: NamedDeclaration): ReadonlyArray<HeritageClause>? {
        val symbol = this.getSymbol(node)
        if (symbol == null) return null

        val heritageClauses = symbol.declarations
            ?.mapNotNull {
                if (!isInterfaceDeclaration(it)) return@mapNotNull null
                it.heritageClauses?.asArray()
            }
            ?.flatMap { it.asIterable() }

        if (isClassDeclaration(node)) {
            val mainHeritageClauses = node.heritageClauses?.asArray()

            return (mainHeritageClauses ?: emptyArray()) +
                    (heritageClauses?.toTypedArray() ?: emptyArray())
        } else {
            return heritageClauses?.toTypedArray() ?: emptyArray()
        }
    }

    fun getMembers(
        node: NamedDeclaration,
        context: Context,
    ): ReadonlyArray<NamedDeclaration>? {
        val symbol = this.getSymbol(node)
        if (symbol == null) return null

        val members = this.getUniqMembers(symbol.members)
            .filter { member -> !isTypeParameterDeclaration(member) }

        val exports = this.getUniqMembers(symbol.exports)
            .filter { member -> !isTypeParameterDeclaration(member) }

        val namespaceInfoService = context.lookupService(namespaceInfoServiceKey)

        return (members + exports).filter { member ->
            val parent = member.parent

            if (
                isModuleBlock(parent)
                && namespaceInfoService?.resolveNamespaceStrategy(parent.parent) == NamespaceStrategy.`package`
            ) {
                return@filter false
            }

            if (isVariableDeclarationList(parent)) {
                val grandparent = parent.parent.parent

                if (
                    isModuleBlock(grandparent)
                    && namespaceInfoService?.resolveNamespaceStrategy(grandparent.parent) == NamespaceStrategy.`package`
                ) {
                    return@filter false
                }
            }

            return@filter true
        }.toTypedArray()
    }

    private fun getUniqMembers(symbolTable: SymbolTable?): ReadonlyArray<NamedDeclaration> {
        val typeChecker = this.program.getTypeChecker()

        return (symbolTable?.let { JsArrays.from(it.values()) } ?: emptyArray())
            .flatMap { symbol ->
                val declarations = symbol.declarations ?: emptyArray()
                val firstDeclaration = declarations.getOrNull(0) ?: return@flatMap emptyList()

                @Suppress("UNCHECKED_CAST_TO_EXTERNAL_INTERFACE")
                if (!isSignatureDeclaration(firstDeclaration)) return@flatMap listOf(firstDeclaration as NamedDeclaration)

                val declarationMap = declarations.fold(DeepMap<Any, NamedDeclaration>()) { acc, declaration ->
                    if (!isSignatureDeclaration(declaration)) return@fold acc

                    val parameterSymbols = declaration.parameters.asArray()
                        .map { parameter ->
                            val typeNode = parameter.type ?: return@map "any"
                            typeChecker.getTypeFromTypeNode(typeNode).symbol.unsafeCast<typescript.Symbol?>()
                                ?: printNode(typeNode)
                        }
                        .toTypedArray()

                    acc[parameterSymbols] = declaration

                    acc
                }

                declarationMap.values().asIterable()
            }
            .toTypedArray()
    }

    @Suppress("CANNOT_CHECK_FOR_EXTERNAL_INTERFACE")
    private fun isSignatureDeclaration(declaration: Declaration): Boolean {
        contract {
            returns(true) implies (declaration is SignatureDeclarationBase)
        }

        return isCallSignatureDeclaration(declaration)
            || isConstructSignatureDeclaration(declaration)
            || isConstructorDeclaration(declaration)
            || isMethodSignature(declaration)
            || isMethodDeclaration(declaration)
    }

    private fun getSymbol(node: NamedDeclaration): typescript.Symbol? {
        val name = node.name ?: return null

        val typeChecker = program.getTypeChecker()
        return typeChecker.getSymbolAtLocation(name)
    }

    private fun printNode(node: Node): String {
        val sourceFile = node.getSourceFileOrNull() ?: virtualSourceFile

        return printer.printNode(EmitHint.Unspecified, node, sourceFile)
    }
}

class DeclarationMergingPlugin(program: Program) : Plugin {
    private val declarationMergingService = DeclarationMergingService(program)

    override fun setup(context: Context) {
        context.registerService(declarationMergingServiceKey, this.declarationMergingService)
    }

    override fun traverse(node: Node, context: Context) = Unit

    override fun render(node: Node, context: Context, next: Render<Node>) = null

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
